filter-and-decorate real-time data ops demo notebook (#1869)
deephaven/deephaven-core: demo/web/src/main/notebooks/A. Real-time table ops/A2 Filter and decorate .md

# Filter and decorate

> Now that we have some data, let's work with it. In this notebook, we show how to decorate and filter our data.

Reviewer: It feels off to start off with "Now that we have some data..." and then immediately make a new table.

> Let's start by creating some daily data. This could simulate something like daily stock prices or daily temperatures.

Reviewer: ```suggestion
Let's start by simulating a year's worth of daily measurements. This could simulate something like stock prices, temperatures, etc.
```

```python
start_time = convertDateTime("2020-01-01T00:00:00 NY")

time_offset = Period("1D")
daily_data = create_random_table(365, start_time, time_offset)
```

> Now we decorate the data by adding its day of the week.<!--TODO what is a good way to describe "decorate"?-->

Reviewer: I think decorate is an acceptable word here

Reviewer 2: agree

Author: I'll remove my comment 🎉

```python
daily_data = daily_data.update("DayOfWeekInt = dayOfWeek(DateTime, TZ_NY)")
```

> And now we convert the day of week to a string representation.

Reviewer 2: ```suggestion
Next,  we convert the day of week to a string representation.
```

```python
import calendar

def day_of_week_int_to_str(day_of_week):
    return calendar.day_name[day_of_week-1]

daily_data = daily_data.update("DayOfWeekStr = day_of_week_int_to_str(DayOfWeekInt)")
```

> Let's do some filtering on this data. For this example, we filter based on values in our columns with simple boolean expressions.

Reviewer: ```suggestion
Deephaven provides a wealth of filtering methods for tables.  We will start by filtering using simple boolean expressions.
```

Reviewer 2: we can link in a markdown notebook, right? link to "how to filter article"

Author: I believe so, I'm not sure about relative links so it may need to be an absolute link (https://deephaven.io/core/docs/how-to-guides/use-filters/)

```python
from deephaven.filter import or_

evens = daily_data.where("Number % 2 == 0")
odds = daily_data.where("!(Number % 2 == 0)")

trues = daily_data.where("Boolean")
falses = daily_data.where("!Boolean")

evens_and_trues = daily_data.where("Number % 2 == 0", "Boolean")
odds_or_falses = daily_data.where(or_("!(Number % 2 == 0)", "!Boolean"))
```

> For this example, we filter based on values in another table.

Reviewer: ```suggestion
Some filtering methods can apply a filter to one table based on another.
```

```python
from deephaven.TableTools import newTable, stringCol

vowels_table = newTable(
    stringCol("Vowels", "A", "E", "I", "O", "U")
)

vowels = daily_data.whereIn(vowels_table, "Character = Vowels")
consonants = daily_data.whereNotIn(vowels_table, "Character = Vowels")
```

> For this example, we filter based on a custom function.

Reviewer: ```suggestion
You can also define custom functions to perform filtering.  Here, the custom function will omit weekend days.
```

> You can also define custom functions to perform filtering. Here, the custom function will omit weekend days.

Reviewer 2: ```suggestion
You can also define custom functions to perform filtering. Here, we omit weekend days.
```
